File: src/robot_driver/robot_driver/MoveForwardActionServer.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse, CancelResponse
from bot_behavior_interfaces.action import MoveForward  # Import the MoveForward action
from geometry_msgs.msg import Twist

class MoveForwardActionServer(Node):
    def __init__(self):
        super().__init__('move_forward_action_server')

        # Declare and set the forward speed directly in the code
        self.forward_speed = 0.2  # Adjust speed as needed

        self._action_server = ActionServer(
            self,
            MoveForward,
            'move_forward',
            self.execute_callback,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback)
        self.publisher_ = self.create_publisher(Twist, 'cmd_vel', 10)
        self.timer = None

    def goal_callback(self, goal_request):
        self.get_logger().info('Received request to move forward.')
        return GoalResponse.ACCEPT

    def cancel_callback(self, goal_handle):
        self.get_logger().info('Received cancel request. Stopping the robot.')
        self.stop_robot()
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle):
        self.get_logger().info(f'Executing move forward with speed {self.forward_speed}...')

        twist_msg = Twist()
        twist_msg.linear.x = self.forward_speed  # Use the class attribute for linear speed

        # Create a timer to publish twist messages periodically
        self.timer = self.create_timer(0.1, self.publish_twist_msg)
        self.timer_msg = twist_msg

        while not goal_handle.is_cancel_requested:
            rclpy.spin_once(self, timeout_sec=0.1)

        if goal_handle.is_cancel_requested:
            self.get_logger().info('Goal canceled. Stopping the robot.')
            self.stop_robot()
            goal_handle.canceled()
            return MoveForward.Result()

    def publish_twist_msg(self):
        if self.timer_msg:
            self.publisher_.publish(self.timer_msg)

    def stop_robot(self):
        # Publish zero velocity to stop the robot
        twist_msg = Twist()
        twist_msg.linear.x = 0.0
        twist_msg.angular.z = 0.0
        self.publisher_.publish(twist_msg)
        if self.timer:
            self.timer.cancel()

def main(args=None):
    rclpy.init(args=args)
    move_forward_action_server = MoveForwardActionServer()
    rclpy.spin(move_forward_action_server)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/robot_driver_cpp/src/robot_driver_cpp/move_forward_action_server.cpp
#include "robot_driver_cpp/move_forward_action_server.hpp"

using namespace std::chrono_literals;

MoveForwardActionServer::MoveForwardActionServer() : Node("move_forward_action_server")
{
    this->declare_parameter<double>("move_speed", 0.2);
    this->get_parameter("move_speed", move_speed_);

    auto options = rclcpp_action::ServerOptions();
    options.result_timeout = std::chrono::milliseconds(0);
    options.goal_handling_timeout = std::chrono::milliseconds(0);
    
    action_server_ = rclcpp_action::create_server<MoveForward>(
        this,
        "move_forward",
        std::bind(&MoveForwardActionServer::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
        std::bind(&MoveForwardActionServer::handle_cancel, this, std::placeholders::_1),
        std::bind(&MoveForwardActionServer::handle_accepted, this, std::placeholders::_1),
        options
    );

    publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);
}

rclcpp_action::GoalResponse MoveForwardActionServer::handle_goal(
    const rclcpp_action::GoalUUID &uuid,
    std::shared_ptr<const MoveForward::Goal> goal
)
{
    RCLCPP_INFO(this->get_logger(), "Received request to move forward.");
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
}

rclcpp_action::CancelResponse MoveForwardActionServer::handle_cancel(
    const std::shared_ptr<GoalHandleMoveForward> goal_handle
)
{
    RCLCPP_INFO(this->get_logger(), "Received cancel request. Stopping the robot.");
    stop_robot();
    return rclcpp_action::CancelResponse::ACCEPT;
}

void MoveForwardActionServer::handle_accepted(const std::shared_ptr<GoalHandleMoveForward> goal_handle)
{
    std::thread{std::bind(&MoveForwardActionServer::execute, this, std::placeholders::_1), goal_handle}.detach();
}

void MoveForwardActionServer::execute(const std::shared_ptr<GoalHandleMoveForward> goal_handle)
{
    RCLCPP_INFO(this->get_logger(), "Executing move forward with speed %f...", move_speed_);
    
    twist_msg_.linear.x = move_speed_;
    timer_ = this->create_wall_timer(
        100ms, [this]() { publish_twist_msg(); }
    );

    while (rclcpp::ok() && !goal_handle->is_canceling()) {
        std::this_thread::sleep_for(100ms);
    }

    if (goal_handle->is_canceling()) {
        RCLCPP_INFO(this->get_logger(), "Goal canceled. Stopping the robot.");
        stop_robot();
        goal_handle->canceled(std::make_shared<MoveForward::Result>());
    } else {
        goal_handle->succeed(std::make_shared<MoveForward::Result>());
    }
}

void MoveForwardActionServer::publish_twist_msg()
{
    if (publisher_) {
        publisher_->publish(twist_msg_);
    }
}

void MoveForwardActionServer::stop_robot()
{
    twist_msg_.linear.x = 0.0;
    twist_msg_.angular.z = 0.0;
    publisher_->publish(twist_msg_);
    if (timer_) {
        timer_->cancel();
    }
}
